tests/test_concurrency.py

```diff
@@ -40,3 +41,43 @@ def endpoint(request: Request) -> Response:
 
     resp = client.get("/")
     assert resp.content == b"data"
+
+
+@pytest.mark.anyio
+async def test_restore_context_from_thread_previously_set():
+    """Value outside of threadpool is overwitten with value set in threadpool"""
+    ctxvar: contextvars.ContextVar[str] = contextvars.ContextVar("ctxvar")
+    ctxvar.set("spam")
+
+    def sync_task():
+        ctxvar.set("ham")
+
+    await run_in_threadpool(sync_task)
+    assert ctxvar.get() == "ham"
+
+
+@pytest.mark.anyio
+async def test_restore_context_from_thread_previously_unset():
+    """Value outside of threadpool is set to value in threadpool"""
+    ctxvar: contextvars.ContextVar[str] = contextvars.ContextVar("ctxvar")
+
+    def sync_task():
+        ctxvar.set("ham")
+
+    await run_in_threadpool(sync_task)
+    assert ctxvar.get() == "ham"
+
+
+@pytest.mark.anyio
+async def test_restore_context_from_thread_new_cvar():
+    """Value outside of threadpool is set for a cvar created in the threadpool"""
+    ctxvars: List[contextvars.ContextVar[str]] = []
+
+    def sync_task():
+        ctxvar: contextvars.ContextVar[str] = contextvars.ContextVar("ctxvar")
+        ctxvar.set("ham")
+        ctxvars.append(ctxvar)
+
+    await run_in_threadpool(sync_task)
+    assert len(ctxvars) == 1
+    assert next(iter(ctxvars)).get() == "ham"
```

Comment:
This are the failing tests, which get to the root of the problem in the PR post without jumping through hoops to exercise the issue